### PR TITLE
[libbeat] Add translate_ldap_attribute processor

### PR DESCRIPTION
**Pull request by marc-gr**
## Proposed commit message

Adds a processor that will connect to an LDAP server in order to translate attributes from one to another. Typically used in Active Directory environments to translate GUID's to their respective common names.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes https://github.com/elastic/integrations/issues/8959

Looking for best approach to add automated tests for this.

### CODE REVIEW

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -111,6 +111,7 @@ https://github.com/elastic/beats/compare/v8.8.1\...main[Check the HEAD diff]
 - Set timeout of 1 minute for FQDN requests {pull}37756[37756]
 - Fix issue where old data could be saved in the memory queue after acknowledgment, increasing memory use {pull}41356[41356]
 - Ensure Elasticsearch output can always recover from network errors {pull}40794[40794]
+- Add `translate_guid` processor for windows platforms. {pull}41472[41472]
```
Comment (mauri870) [inline CHANGELOG.next.asciidoc:114]: Shouldn't this processor include `ldap` somewhere in its name, for clarity?

Comment (marc-gr) [reply CHANGELOG.next.asciidoc:114]: Agreed, initally was expected to be less generic.

File: filebeat/tests/integration/translate_guid_test.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	reader, err := c.ImagePull(ctx, "bitnami/openldap:2", image.PullOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.Copy(os.Stdout, reader)
+	reader.Close()
+
+	resp, err := c.ContainerCreate(ctx,
+		&container.Config{
+			Image: "bitnami/openldap:2",
```
Comment (mauri870) [inline filebeat/tests/integration/translate_guid_test.go:146]: I'm not that familiar with LDAP, but here it is installing a 2.x openldap and the go client is v3. I wonder if these versions should match?

Comment (marc-gr) [reply filebeat/tests/integration/translate_guid_test.go:146]: The ldap library refers to the library version, openldap version is just the version of openldap server in this case. They are not correlated

Comment (mauri870) [reply filebeat/tests/integration/translate_guid_test.go:146]: Got it, thanks!

File: libbeat/processors/translate_ldap_attribute/docs/translate_ldap_attribute.asciidoc [hunk 3]
```diff
@@ -0,0 +1,92 @@
+[[processor-translate-guid]]
+=== Translate GUID
+
+++++
+<titleabbrev>translate_ldap_attribute</titleabbrev>
+++++
+
+The `translate_ldap_attribute` processor translates an LDAP attributes between eachother.
```
Comment (kilfoyle) [inline libbeat/processors/translate_ldap_attribute/docs/translate_ldap_attribute.asciidoc:8]:
```suggestion
The `translate_ldap_attribute` processor translates LDAP attributes between entities.
```
I'm not sure if "entities" is the right word. Just a guess on my part. :-)

**Review by kilfoyle — APPROVED** [review]
LGTM for the Asciidoc part. 🚀 
Just one super small suggestion.

File: filebeat/tests/integration/translate_guid_test.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+	c, err := docker.NewClient(client.DefaultDockerHost, nil, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	reader, err := c.ImagePull(ctx, "bitnami/openldap:2", image.PullOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.Copy(os.Stdout, reader)
```
Comment (mauri870) [inline filebeat/tests/integration/translate_guid_test.go:141]: Can you add a check for err here like the linter is suggesting? Thanks.

File: libbeat/processors/translate_ldap_attribute/ldap.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	defer client.mu.Unlock()
+
+	// Connect with or without TLS based on configuration
+	var opts []ldap.DialOpt
+	if client.tlsConfig != nil {
+		opts = append(opts, ldap.DialWithTLSConfig(client.tlsConfig))
+	}
+	conn, err := ldap.DialURL(client.address, opts...)
+	if err != nil {
+		return fmt.Errorf("failed to dial LDAP server: %v", err)
```
Comment (mauri870) [inline libbeat/processors/translate_ldap_attribute/ldap.go:70]: %w is a better fit here

File: libbeat/processors/translate_ldap_attribute/ldap.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+	if client.password != "" {
+		err = conn.Bind(client.username, client.password)
+	} else {
+		err = conn.UnauthenticatedBind(client.username)
+	}
+
+	if err != nil {
+		conn.Close()
+		return fmt.Errorf("failed to bind to LDAP server: %v", err)
```
Comment (mauri870) [inline libbeat/processors/translate_ldap_attribute/ldap.go:81]: %w here as well

File: libbeat/processors/translate_ldap_attribute/ldap.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+		return client.connect()
+	}
+	return nil
+}
+
+// findObjectBy searches for an object and returns its mapped values.
+func (client *ldapClient) findObjectBy(searchBy string) ([]string, error) {
+	// Ensure the connection is alive or reconnect if necessary
+	if err := client.reconnect(); err != nil {
+		return nil, fmt.Errorf("failed to reconnect: %v", err)
```
Comment (mauri870) [inline libbeat/processors/translate_ldap_attribute/ldap.go:104]: %w here too

**Comment by cmacknz** [comment]
Update the CODEOWNERS to have the security windows platform team be the explicitly owners of the new processor please :)

File: libbeat/processors/translate_ldap_attribute/docs/translate_ldap_attribute.asciidoc (hunk 3 above)
Comment (marc-gr) [reply libbeat/processors/translate_ldap_attribute/docs/translate_ldap_attribute.asciidoc:8]: It translates an attribute of an entity to another one of the same entity eg Get a Name based on an ID

**Review by mauri870 — APPROVED** [review]

**Review by VihasMakwana — APPROVED** [review]

**Review by cmacknz — APPROVED** [review]
Approving codeowner change

**Pull request merged**